File: mlua-gen/tests/on_set_enum.lua
assert(hits() == 0)

-- Setter for an unnamed variant.
state.active = { 5 }
assert(hits() == 1)
assert(state.active[1] == 5)
assert(hits() == 1)

-- Setter for a named variant.
state.labelled = { name = "x", count = 1 }
assert(hits() == 2)
assert(state.labelled.name == "x")
assert(state.labelled.count == 1)
assert(hits() == 2)

-- In-place mutation through the variant proxy fires `on_set` once per
-- top-level `=` (Phase 3).
state.labelled.name = "y"
assert(hits() == 3)
assert(state.labelled.name == "y")
assert(hits() == 3)

-- Switching variant first, then in-place mutation.
state.active = { 10 }
assert(hits() == 4)
state.active[1] = 20
assert(hits() == 5)
assert(state.active[1] == 20)

File: mlua-gen/tests/on_set_enum.rs
//! `on_set = <path>` on an enum: the hook fires once per Lua-side variant
//! swap, for both unnamed and named variants.

use {
    mlua_gen::{LuaBuilder, mlua_gen},
    std::sync::atomic::{AtomicUsize, Ordering},
};

static HITS: AtomicUsize = AtomicUsize::new(0);

fn on_set_hook() {
    HITS.fetch_add(1, Ordering::Relaxed);
}

#[mlua_gen(on_set = crate::on_set_hook)]
#[allow(dead_code)] // variants are reachable through Lua, not Rust
enum State {
    Idle,
    Active(u32),
    Labelled { name: String, count: u32 },
}

#[test]
pub fn test() {
    let lua = mlua::Lua::new();
    State::to_globals(&lua).unwrap();

    // Seed with an Idle value so the .lua side can mutate it through the
    // generated variant setters.
    lua.globals().set("state", State::Idle).unwrap();

    lua.globals()
        .set(
            "hits",
            lua.create_function(|_, ()| Ok(HITS.load(Ordering::Relaxed)))
                .unwrap(),
        )
        .unwrap();

    lua.load(include_str!("./on_set_enum.lua")).exec().unwrap();
}
